### PR TITLE
Add 'udevadm control --ping'

### PR DESCRIPTION
**Pull request by yuwata**
When udevd is started from sysv init script, then this may be useful to check udevd is started or not.

c.f. https://github.com/systemd/systemd/issues/11314#issuecomment-451318393.

cc @mbiebl

### CODE REVIEW

**Comment by yuwata** [comment]
Or, instead of this, `udevadm trigger` may optionally ping before triggering uevents.

**Comment by yuwata** [comment]
The '--ping' option is also added to 'trigger' command.

**Comment by yuwata** [comment]
Also added a similar mechanism in udevd.

**Comment by yuwata** [comment]
Updated.
- `timeout` in `udev_ctrl_send_*()` is changed from `int` to `usec_t`.

File: src/udev/udevadm-control.c [hunk 1]
```diff
@@ -77,7 +77,7 @@ int control_main(int argc, char *argv[], void *userdata) {
         }
 
         if (argc <= 1)
-                log_error("Option missing");
+                return help();
```
Comment (keszybz) [inline src/udev/udevadm-control.c]: No, please don't do this. Printing the full help is just not helpful, if we know what exactly is wrong. We don't do this anywhere else in the code base.

File: src/udev/udevd.c [hunk 2]
```diff
@@ -1035,6 +1035,7 @@ static int on_ctrl_msg(sd_event_source *s, int fd, uint32_t revents, void *userd
         if (i >= 0) {
                 log_debug("Received udev control message (SET_LOG_LEVEL), setting log_priority=%i", i);
                 log_set_max_level(i);
+                log_set_max_level_realm(LOG_REALM_SYSTEMD, i);
```
Comment (keszybz) [inline src/udev/udevd.c:1031]: Hmm, can you describe (in the commit message) what the effect of this is? I think I understand, but it's easier to review the patch if the *intent* is clear.

File: src/udev/udevadm-control.c [hunk 3]
```diff
@@ -135,6 +141,11 @@ int control_main(int argc, char *argv[], void *userdata) {
                                 return r;
                         break;
                 }
+                case ARG_PING:
+                        r = udev_ctrl_send_ping(uctrl, timeout);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to connect to udev daemon.");
```
Comment (keszybz) [inline src/udev/udevadm-control.c]: ": %m" ?

File: man/systemd-udevd.service.xml [hunk 4]
```diff
@@ -121,6 +121,17 @@
...
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--ping-timeout=</option></term>
+        <listitem>
+          <para>Set the number of seconds to wait for daemon to finish initialization.
+          if 0 is specified, then the parent process does not wait and exit immediately
+          after the daemon process is forked.
+          If <option>--daemon</option> is not specified, the specified value is ignored.
+          The default is 5 seconds.</para>
```
Comment (keszybz) [inline man/systemd-udevd.service.xml]: I think this should be caclled "--daemon-timeout"? It then matches the fact that "--daemon" option is used nicely.

File: src/udev/udevd.c [hunk 5]
```diff
@@ -1848,9 +1868,22 @@ static int run(int argc, char *argv[]) {
...
+                        if (arg_ping_timeout_usec == 0)
+                                return 0;
+
+                        uctrl = udev_ctrl_new();
+                        if (!uctrl)
+                                return log_oom();
+
+                        r = udev_ctrl_send_ping(uctrl, arg_ping_timeout_usec);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to connect to udev daemon.");
```
Comment (keszybz) [inline src/udev/udevd.c]: ": %m" ?

File: src/udev/udevadm-trigger.c [hunk 6]
```diff
@@ -286,6 +306,24 @@ int trigger_main(int argc, char *argv[], void *userdata) {
...
+                if (r < 0)
+                        return r;
+        }
+
+        if (ping) {
+                _cleanup_(udev_ctrl_unrefp) struct udev_ctrl *uctrl = NULL;
+
+                uctrl = udev_ctrl_new();
+                if (!uctrl)
+                        return -ENOMEM;
```
Comment (keszybz) [inline src/udev/udevadm-trigger.c]: `log_oom()` ?

File: src/udev/udevadm-control.c [hunk 7]
```diff
@@ -40,6 +40,7 @@ static int help(void) {
                "  -R --reload              Reload rules and databases\n"
                "  -p --property=KEY=VALUE  Set a global property for all events\n"
                "  -m --children-max=N      Maximum number of children\n"
+               "     --ping                Send a ping message\n"
```
Comment (keszybz) [inline src/udev/udevadm-control.c]: Maybe describe this as "Wait for udev to respond to a ping message"

File: man/udevadm.xml [hunk 8]
```diff
@@ -419,6 +419,12 @@
             same time.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><option>--ping</option></term>
+          <listitem>
+            <para>Send a ping message.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: I think this needs to mention that it waits for the daemon to respond, with a configurable timeout.

File: man/udevadm.xml [hunk 9]
```diff
@@ -319,6 +319,19 @@
             the same command to finish.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><option>--ping</option></term>
+          <listitem>
+            <para>Before triggering events, check that <command>systemd-udevd</command> daemon is already started.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: I think it's better not to add two options for this. Just add `--timeout=` or `--daemon-timeout=`, and if it is specified, wait the specified time. The fact that this is implemented as a ping message is just an internal naming detail. The description of the option needs to say that this waits for the daemon to start.

File: src/udev/udevadm-control.c (hunk 1 above)
Comment (yuwata) [reply src/udev/udevadm-control.c]: OK. I will drop that part.
> We don't do this anywhere else in the code base.

cryptsetup and veritysetup do that. But, yeah, normally they are not run by user.

File: src/udev/udevd.c (hunk 2 above)
Comment (yuwata) [reply src/udev/udevd.c:1031]: OK. I will update the commit message. Sorry for my laziness.

The reason is, now not a few `udevd` debug logs are come from sd-device or sd-hwdb. Only setting `LOG_REALM_UDEV` may not sufficient to debug. We already have similar code in `main()` and `udevadm`.

File: src/udev/udevadm-control.c (hunk 7 above)
Comment (yuwata) [reply src/udev/udevadm-control.c]: Thanks.

**Comment by mbiebl** [comment]
Hm,...
```
# /lib/sytemd/systemd-udevd --daemon
starting version 240-135-g5cefefc
# echo $?
1
```
The init script and the udev initramfs hooks both use `set -e` and fail at this point.
This would explain the CI failures.

**Comment by mbiebl** [comment]
I also noticed this:
screenshot_debian_buster_2019-01-09_22 14 49

File: src/shared/udev-util.c [hunk 10]
```diff
@@ -89,6 +92,12 @@ int udev_parse_config_full(
                         *ret_resolve_name_timing = t;
         }
 
+        if (ret_ping_timeout_usec && ping_timeout) {
+                r = parse_sec(ping_timeout, ret_ping_timeout_usec);
+                if (r < 0)
+                        log_notice_errno(r, "/etc/udev/udev.conf: failed to set parse ping_timeout=%s, ignoring: %m", ping_timeout);
```
Comment (mbiebl) [inline src/shared/udev-util.c]: `failed to set parse` → either set or parse?

File: src/udev/udev.conf [hunk 11]
```diff
@@ -6,5 +6,6 @@
 #udev_log=info
 #children_max=
 #exec_delay=
-#event_timeout=180
+#event_timeout=180s
```
Comment (mbiebl) [inline src/udev/udev.conf]: Is this backwards compatible, i.e. if someone has used `event_timeout=90` will that continue to work?

File: src/udev/udevd.c (hunk 5 above)
Comment (mbiebl) [inline src/udev/udevd.c]: seems I'm hitting this error condition, see the screenshot at https://github.com/systemd/systemd/pull/11349#issuecomment-452858706

**Comment by yuwata** [comment]
Updated. `--ping` option or friends are dropped, as it is not necessary anymore.

**Comment by mbiebl** [comment]
@yuwata I think the `--ping` command might still be useful, at least for Debian's sysv init script.
We reliably need to determine the PID of the udevd process. Unfortunately, udevd no longer writes a PID file itself in daemon mode.
So we decided to not use `--daemon` but instead run it in the foreground. But then we have the problem, that we don't know when udevd is ready to process requests, i.e. when it is safe to call `udevadm trigger`

Maybe you have clever idea how we can either
a/ use `--daemon` mode and reliably determine the PID of the systemd-udevd process
or
b/ use foreground mode and reliably determine when udev is ready to process requests

**Comment by yuwata** [comment]
@mbiebl Sorry, I forgot `start-stop-daemon` case. The option '--ping' is re-added. PTAL.

File: src/udev/udevadm-control.c [hunk 12]
```diff
@@ -40,6 +40,7 @@ static int help(void) {
                "  -R --reload              Reload rules and databases\n"
                "  -p --property=KEY=VALUE  Set a global property for all events\n"
                "  -m --children-max=N      Maximum number of children\n"
+               "     --ping                Wait for udev to respond to a ping message"
```
Comment (keszybz) [inline src/udev/udevadm-control.c]: Missing `\n`.

Comment (keszybz) [reply src/udev/udevadm-control.c]: I force-pushed to add new newline.

**Comment by keszybz** [comment]
It was passing before my cosmetic repush, so let's assume it's OK with half of the CI passing.

**Comment by keszybz** [comment]
`E: Malformed entry 1 in list file /etc/apt/sources.list.d/autopkgtest.list (URI parse)`
for all bionic-* instances.

File: src/udev/udevadm-control.c (hunk 12 above)
Comment (yuwata) [reply src/udev/udevadm-control.c]: Thanks.

**Pull request merged**